Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+Factory.swift

```diff
@@ -0,0 +1,314 @@
+        }
+    }
+}
+
+extension HTTPConnectionPool.ConnectionFactory {
+    func makeBestChannel(connectionID: HTTPConnectionPool.Connection.ID, eventLoop: EventLoop, logger: Logger) -> EventLoopFuture<(Channel, HTTPVersion)> {
+        if self.key.scheme.isProxyable, let proxy = self.clientConfiguration.proxy {
+            return self.makeHTTPProxyChannel(proxy, connectionID: connectionID, eventLoop: eventLoop, logger: logger)
+        } else {
+            return self.makeChannel(eventLoop: eventLoop, logger: logger)
+        }
+    }
+
+    private func makeChannel(eventLoop: EventLoop, logger: Logger) -> EventLoopFuture<(Channel, HTTPVersion)> {
+        switch self.key.scheme {
+        case .http, .http_unix, .unix:
+            return self.makePlainChannel(eventLoop: eventLoop).map { ($0, .http1_1) }
+        case .https, .https_unix:
+            return self.makeTLSChannel(eventLoop: eventLoop, logger: logger).map {
+                (channel, negotiated) -> (Channel, HTTPVersion) in
+                let version = negotiated == "h2" ? HTTPVersion.http2 : HTTPVersion.http1_1
+                return (channel, version)
+            }
+        }
+    }
+
+    private func makePlainChannel(eventLoop: EventLoop) -> EventLoopFuture<Channel> {
+        let bootstrap = self.makePlainBootstrap(eventLoop: eventLoop)
+
+        switch self.key.scheme {
+        case .http:
+            return bootstrap.connect(host: self.key.host, port: self.key.port)
+        case .http_unix, .unix:
+            return bootstrap.connect(unixDomainSocketPath: self.key.unixPath)
+        case .https, .https_unix:
+            preconditionFailure("Unexpected schema")
+        }
+    }
+
+    private func makeHTTPProxyChannel(
+        _ proxy: HTTPClient.Configuration.Proxy,
+        connectionID: HTTPConnectionPool.Connection.ID,
+        eventLoop: EventLoop,
+        logger: Logger
+    ) -> EventLoopFuture<(Channel, HTTPVersion)> {
+        // A proxy connection starts with a plain text connection to the proxy server. After
+        // the connection has been established with the proxy server, the connection might be
+        // upgraded to TLS before we send our first request.
+        let bootstrap = self.makePlainBootstrap(eventLoop: eventLoop)
+        return bootstrap.connect(host: proxy.host, port: proxy.port).flatMap { channel in
+            let connectPromise = channel.eventLoop.makePromise(of: Void.self)
+
+            let encoder = HTTPRequestEncoder()
+            let decoder = ByteToMessageHandler(HTTPResponseDecoder(leftOverBytesStrategy: .dropBytes))
+            let proxyHandler = HTTP1ProxyConnectHandler(
+                targetHost: self.key.host,
+                targetPort: self.key.port,
+                proxyAuthorization: proxy.authorization,
+                connectPromise: connectPromise
+            )
+
+            do {
+                try channel.pipeline.syncOperations.addHandler(encoder)
```

Comment:
why not add all handler at once?

Reply:
+1

Reply:
Because this way we would need to alloc an array. And if possible, we want to save this alloc.

Reply:
@Lukasa any opinion?

Reply:
I'm happy to do these one at a time, as-is currently implemented.